### DISCUSSION TITLE
bibtex with mkdocstrings?

### DISCUSSION THREAD

**Discussion by bobmyhill** [mkdocstrings/mkdocstrings#482] [Q&A] [answered]
I'm using mkdocstrings with sphinx-style markdown, and would like to add references to published papers with bibtex.
I've tried mkdocs-bibtex. The recommended formatting (`[@reference]`) works fine within normal markdown, but neither `[@reference]` nor the sphinx-style `:cite:` formatting works for mkdocstrings autodoc. 

Is there something I'm missing / a workaround?
Thanks!

**Comment by pawamoy** [comment]
Sphinx directives (`:cite:`) will definitely not work. Do you have a public repository that I can use to replicate the issue and investigate?

**Comment by bobmyhill** [comment]
There is! It's a niche project, but I'd like to expand my use of mkdocstrings to other mainstream codes:
https://github.com/mantle-convection-constrained/terratools

At the moment, you'll see places in the code where I use [reference], and the references md then includes all the [@......]s. It would be great to replace the former with the latter in the docstrings.

Thank you for offering to investigate!

**Comment by oprypin** [comment]
> I've tried mkdocs-bibtex. The recommended formatting (`[@reference]`) works fine within normal markdown, but neither `[@reference]` nor the sphinx-style `:cite:` formatting works for mkdocstrings autodoc.

How are you trying it?
I suspect if you add the reference definitions within the docstring itself, then it will work. But of course it's not ideal.
But have you tried putting the references into an auto-append file?
https://squidfunk.github.io/mkdocs-material/reference/tooltips/#adding-a-glossary

Ah or actually I was simply looking at https://github.com/darwindarak/mdx_bib - my answer applies to that one. Which one are you using?

**Comment by oprypin** [comment]
In case you're referring to https://github.com/shyamd/mkdocs-bibtex then yes, I can see why it wouldn't work inside docstrings.
It looks at the raw page Markdown, and indeed in that case there's no way for it to process dynamically inserted content from mkdocstrings.
https://github.com/shyamd/mkdocs-bibtex/blob/184618fd2b2cb872002f174d64cb13957ba1b36a/src/mkdocs_bibtex/plugin.py#L95
Instead they should've done it as a Python-Markdown extension, like https://github.com/darwindarak/mdx_bib does. Maybe you can try that one out instead, as per my other answer.

**Reply by bobmyhill** [reply]
Thanks! Yes, I was indeed using the https://github.com/shyamd/mkdocs-bibtex plugin (via `pip install`).
I've tried using https://github.com/darwindarak/mdx_bib, but I get a config error "Config value: 'markdown_extensions'. Error: Failed to load extension  'mdx_bib'".

I get the same error when running `mkdocs serve` with this package: https://github.com/tramebleue/fct-qgis after installing all the required modules in `requirements.txt` via pip.

I assume this is a versioning / deprecation issue, but I can't work out where the problem is. Any suggestions / advice gratefully received.

**Reply by bobmyhill** [reply]
I've just spent an hour playing with https://github.com/darwindarak/mdx_bib, but I can't get it to work satisfactorily, nor can I find any projects where it is used successfully. 

The `extendMarkdown` function in that package is where I see failures, specifically:
```
        md.preprocessors.add("mdx_bib", CitationsPreprocessor(self.bib), "<reference")
        md.inlinePatterns.add(
            "mdx_bib", CitationsPattern(CITATION_RE, self.bib), "<reference"
        )
        md.treeprocessors.add("mdx_bib", CitationsTreeprocessor(self.bib), "_begin")
```
The argument "<reference" throws a value error from within markdown utils: ValueError: No item named "reference" exists.
Shrug. I guess this isn't a mkdocstrings issue so much as the fact there isn't a fully functional bib extension.

Thanks for your help, and any other suggestions gratefully received!

**Reply by oprypin** [reply]
@bobmyhill 
OK, sorry to lead you on this wild goose chase.
I see that that that extension broke since Markdown 3.3 due to [this specific change](https://github.com/python-markdown/markdown/commit/b701c34ebd7b2d0eb319517b9a275ddf0c89608d#diff-b2bf5edde544f69330a763b3ca9158cfd272544b2727af45d8dd2ca58d464a1bL37).
But I can't simply tell you to install that old version of Markdown because then other things break.

So instead I fix this extension for modern versions :)

<pre>
pip install git+<a href="https://github.com/oprypin/mdx_bib">https://github.com/oprypin/mdx_bib</a>
</pre>

Now, I still can't be sure that it actually works because I don't have a bibtex example.

So, you can try it out yourself, likely with the hint that's in my other answer, or give me an example of what references you want to apply.

**Reply by bobmyhill** [reply]
@oprypin 
Wow, thank you for updating the module for me :) ... I spent some time looking for the incompatibilities myself after sending you the message above, but you're clearly much better at it!

I've used your new version to make the documentation, and `mkdocs serve` now runs.
The (hopefully one) issue remaining is that the references appear multiple times on a single documentation page:
https://terratools.readthedocs.io/en/test_bib/reference/terratools/properties/attenuation/

The branch is here: https://github.com/mantle-convection-constrained/terratools/tree/test_bib

**Reply by oprypin** [reply]
I intend to keep looking into this but haven't had time yet.
Actually at first guess it "makes sense" why it's happening - we're calling Markdown several times in mkdocstrings, and the extension produces the list of references as many times :/

**Reply by bobmyhill** [reply]
Thanks! I've fixed most of the problems in mdxbib, will post my solution in an hour or two.

**Comment by bobmyhill** [comment, marked as answer]
@oprypin Thank you so much for your help. I've now made some changes of my own to mdx_bib (https://github.com/bobmyhill/mdx_bib), and everything now works nicely.
